File: src/algorithms/matrices/transformation/flip-image/sources/flip-image.ts
// Flip and Invert Binary Image (LeetCode 832)
// Flip each row horizontally (reverse), then invert every element (0→1, 1→0).
// Time: O(m × n) — each element touched once
// Space: O(1) — in-place

function flipImage(matrix: number[][]): number[][] {
  const rowCount = matrix.length; // @step:initialize
  const colCount = matrix[0]?.length ?? 0; // @step:initialize

  for (let rowIdx = 0; rowIdx < rowCount; rowIdx++) {
    let leftCol = 0; // @step:flip-cell
    let rightCol = colCount - 1; // @step:flip-cell

    // Two-pointer: swap and XOR-invert simultaneously from both ends
    while (leftCol < rightCol) {
      const leftVal = matrix[rowIdx]![leftCol]!; // @step:flip-cell
      const rightVal = matrix[rowIdx]![rightCol]!; // @step:flip-cell
      matrix[rowIdx]![leftCol] = rightVal ^ 1; // @step:flip-cell
      matrix[rowIdx]![rightCol] = leftVal ^ 1; // @step:flip-cell
      leftCol++; // @step:flip-cell
      rightCol--; // @step:flip-cell
    }

    // When colCount is odd, middle element only needs inversion
    if (leftCol === rightCol) {
      matrix[rowIdx]![leftCol] = matrix[rowIdx]![leftCol]! ^ 1; // @step:flip-cell
    }
  }

  return matrix; // @step:complete
}

File: src/algorithms/matrices/transformation/flip-image/sources/flip-image.rs
// Flip and Invert Binary Image (LeetCode 832)
// Flip each row horizontally (reverse), then invert every element (0→1, 1→0).
// Time: O(m × n) — each element touched once
// Space: O(1) — in-place

fn flip_image(matrix: &mut Vec<Vec<i32>>) -> &Vec<Vec<i32>> {
    let row_count = matrix.len(); // @step:initialize
    let col_count = if row_count > 0 { matrix[0].len() } else { 0 }; // @step:initialize

    for row_idx in 0..row_count {
        let mut left_col: i32 = 0; // @step:flip-cell
        let mut right_col: i32 = (col_count - 1) as i32; // @step:flip-cell

        // Two-pointer: swap and XOR-invert simultaneously from both ends
        while left_col < right_col {
            let left_val = matrix[row_idx][left_col as usize]; // @step:flip-cell
            let right_val = matrix[row_idx][right_col as usize]; // @step:flip-cell
            matrix[row_idx][left_col as usize] = right_val ^ 1; // @step:flip-cell
            matrix[row_idx][right_col as usize] = left_val ^ 1; // @step:flip-cell
            left_col += 1; // @step:flip-cell
            right_col -= 1; // @step:flip-cell
        }

        // When col_count is odd, middle element only needs inversion
        if left_col == right_col {
            matrix[row_idx][left_col as usize] ^= 1; // @step:flip-cell
        }
    }

    matrix // @step:complete
}
